File: recursion/python/solutions/reverse.py
"""Solution for reverse."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list
from append import append


def reverse(lst):
    """
    Return a new list with elements in reverse order.

    reverse(1 -> 2 -> 3) => 3 -> 2 -> 1
    reverse(()) => ()
    """
    if is_empty(lst):
        return EMPTY_LIST

    first, rest = unprepend(lst)

    return append(first, reverse(rest))


if __name__ == '__main__':
    # Test cases
    assert to_python_list(reverse(Node.from_list([1, 2, 3]))) == [3, 2, 1]
    assert to_python_list(reverse(EMPTY_LIST)) == []
    assert to_python_list(reverse(Node.from_list([42]))) == [42]
    assert to_python_list(reverse(Node.from_list(['a', 'b', 'c']))) == ['c', 'b', 'a']

    print("All tests passed!")

File: recursion/javascript/solutions/reverse.js
/**
 * Solution for reverse.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

// Helper: append to end of list
function append(val, list) {
  if (isEmpty(list)) {
    return prepend(val, EMPTY_LIST);
  }

  let [first, rest] = unprepend(list);

  return prepend(first, append(val, rest));
}

function reverse(list) {
  if (isEmpty(list)) {
    return EMPTY_LIST;
  }

  let [first, rest] = unprepend(list);

  return append(first, reverse(rest));
}

if (require.main === module) {
  console.assert(JSON.stringify(toArray(reverse(Node.fromArray([1, 2, 3])))) === '[3,2,1]');
  console.assert(JSON.stringify(toArray(reverse(Node.fromArray([42])))) === '[42]');
  console.assert(JSON.stringify(toArray(reverse(EMPTY_LIST))) === '[]');
  console.assert(JSON.stringify(toArray(reverse(Node.fromArray([1, 2, 3, 4, 5])))) === '[5,4,3,2,1]');

  console.log('All tests passed!');
}

module.exports = { reverse, append };
